File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n=s.size();
        unordered_map<char,int>mp1;
        int j=0;
        int maxlength=0;
        for(int i=0;i<n;i++){
            while(mp1.find(s[i])!=mp1.end()){
               mp1[s[j]]--;
               if(mp1[s[j]]==0){
                   mp1.erase(s[j]);
               }
               j++;
            }
               mp1[s[i]]++;
               maxlength=max(maxlength,i-j+1);
        }
        return maxlength;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        HashMap<Character,Integer>mp = new HashMap<>();
        int j=0;
        int length=0;
        for(int i=0;i<n;i++){
            char ch = s.charAt(i);
            while(mp.containsKey(ch)){
                mp.put(s.charAt(j),mp.getOrDefault(s.charAt(j),0)-1);
                if(mp.get(s.charAt(j))==0){
                    mp.remove(s.charAt(j));
                }
                j++;
            }
            length=Math.max(length,i-j+1);
            mp.put(ch,mp.getOrDefault(ch,0)+1);
        }
        return length;
    }
}